Refactor test and coverage logic

dupdob opened the pull request:
This PR improves on the existing test and coverage logic on several front:

- [x] **Drastically reduce the memory pressure**  during testing phase thanks to the suppression of several wrapping classes. This results in a slight performance increase for large projects
- [x] **Improve support for dynamic test cases** (i.e. using `TestCaseSource` attribute) thanks to extensive testing of xUnit and NUnit behaviour. This should fixes several issues related to those as well as increase Stryker accuracy and efficiency.
- [x] **Improve support for test cases providing multiple results**. This should fix potential bus with those (mostly false negative, i.e. mutants flagged as survivor while they should be killed by some test).
- [x] **Improved timeout computation** based on the list of tests to executes (vs reusing the default test time as before). The current logic is: fix/initialization time + sum of the run time of the tests (as established during initial test run), which then goes through the computation logic. the fix/initialization time is estimated by subtracting the sum of test times to the initial run time. 
- [x] **tolerate failing tests** in the initial test run (i.e. with the non mutated assembly).

see #1476 
Other changes and details
- remove `DotNetTestRunner` code
- remove named pipes and related classes
- tests failing at initial run:
  -  Stryker now tolerates up to 10% of failed tests during initial test run. This is an arbitrary decision that can easily be changed, but I felt having a threshold was sensible
  - _'failing tests'_ are still run during mutation tests and **considered as successful**
  - corollary : a mutant covered only by _'failing tests'_ is flagged as _Survived_. It could have been flagged as 'No Coverage' but I felt user will be surprised by this
- I refactored the backup/restore of the assembly so that the same class (`ProjectInfo`) was in charge of both backup AND restore
- the `MutationInput` class now hosts an `ITimeoutValueCalculator' and the results of the initial Testrun. Several method signatures, interfaces and test have been refactored as a consequence
- the `Mutant` class no longer stores complete description of the covering tests, only the test ids, for efficiency reason. As a consequence, I had to slight change the MutantFilter class so the `DiffMutantFilter` still support changed test source files
- several tests have been added to cover testcases with multiple test results
- Stryker now detects when mutation are executed outside normal test run (usually before the test case starts). Those mutations are marked to be run against all tests as a precaution; we may adopt a finer strategy at later time
- Stryker's data collector set the correct mutant ID when each test starts, except when there is only one mutant to test in the while session. Then the mutation ID is set at the beginning of the test session. This is required for mutation appearing outside tests
- I also fixed my issues and some cosmetic refactoring as suggested by Resharper

jnyrup commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+   xUnit runner calls data theory's data source to feach each test value(s)
+TestCase start event
+   xUnit run test with first set of data
+   xUnit reports test reult
```
reult -> result/results

jnyrup commented on docs/technical-reference/Testing Framework.md:
````diff
@@ -0,0 +1,114 @@
+
+Here is a summarized timeline of tests execution:
+```
+   xUnit runner calls data theory's data source to feach each test value(s)
````
feach -> fetch

jnyrup commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+
+## Generalities
+The [VsTest protocol](https://github.com/Microsoft/vstest-docs/blob/master/RFCs/0006-DataCollection-Protocol.md) is not
+rich enought to cover evey test framework features. Therefore, each framework makes their own decision on how to handle
```
```suggestion
rich enough to cover every test framework feature. Therefore, each framework makes their own decisions on how to handle
```

jnyrup commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+rich enought to cover evey test framework features. Therefore, each framework makes their own decision on how to handle
+some of their specificities. That is the reason why we can observe different behaviour and surprising results.
+
+This affect:
```
```suggestion
This affects:
```

jnyrup commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+- how test results are reported
+
+## Identification
+By default, VsTest generates unique identifier stored as `Guid`. I do not know if test runners are able to provide
```
"unique identifier" -> "a unique identifier" or "unique identifiers"

jnyrup commented on docs/technical-reference/Testing Framework.md:
````diff
@@ -0,0 +1,114 @@
+TestCase start event
+   ...
+```
+The difficuly here is that a lot happens between 'testcase end' and 'testcase start' event. 
````
difficuly -> difficulty

jnyrup commented on docs/technical-reference/Testing Framework.md:
````diff
@@ -0,0 +1,114 @@
+```
+The difficuly here is that a lot happens between 'testcase end' and 'testcase start' event. 
+At coverage capture it is a problem, because there is a risk of spilling coverage information to the next tests.
+And during execution phase, it is diffcult to predict when the test will really be over. 
````
diffcult -> difficult

jnyrup commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+
+### TestCases
+NUnit's TestCases is a pattern where some test is executed multiple time with varying input data (like xUnit's TestCase).
+These test mesthods accept one or more parameters and bear one of this attributes (on top of the `[Test]` attribue):
```
```suggestion
These test mesthods accept one or more parameters and bear one of these attributes (on top of the `[Test]` attribute):
```

jnyrup commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+
+The problems for Stryker are:
+1. If test cases have the same name, we cannot distinguish between test cases, so it must be handled as a single test. This *single test* reports multiple test cases so Stryker needs to wait either for one failure or for each testcase
+to succeed to establish the correct status. The good news is that discovery should provide the number of result to wait for.
```
```suggestion
to succeed to establish the correct status. The good news is that discovery should provide the number of results to wait for.
```

jnyrup commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+The problems for Stryker are:
+1. If test cases have the same name, we cannot distinguish between test cases, so it must be handled as a single test. This *single test* reports multiple test cases so Stryker needs to wait either for one failure or for each testcase
+to succeed to establish the correct status. The good news is that discovery should provide the number of result to wait for.
+2. Every mutations encountered by the `TestCaseSource` method will be reported as covered by the first test case,
```
"Every mutations" -> "every mutation"/"all mutations"

jnyrup reviewed (commented):
Will this handle MSTestV2?

dupdob commented:
thanks for spotting my typos. These changes have no impact on MsTest support. It should work as before; I had easy access to xUnit and NUnit test projects, so I did extensive testing analysis for those.

dupdob commented:
TLDR; because it was the pragmatic choice.
Globally, we want failing tests to have **no impact on mutation test results**.
If failing tests are considered as:
- _Failing_ : they will kill mutants!!
- _Skipped_: some mutants will stay in not run status, because some covering tests are not listed as ran.
- _Ignored_: if a mutation is only covered by failing tests, it will be marked as `NoCoverage`, which will not be understood by user.

The best approach would be to add some new state for those tests and implement proper filtering logic, but it provides no value whatsoever as of now.

rouke-broersma commented:
I see. We should probably plan that improvement when we pick up #1565 as then we can provide better insights into what the tests did (and that they were failing from the beginning, which resulted in mutants having status x)

dupdob commented:
Hi, as far as I understand the situation, there is nothing left to discuss on this PR, or is there any topics that blocks its merge

rouke-broersma commented:
Hi @dupdob I haven't had time to review so that's why I haven't merged

dupdob commented:
No problem, I now this is a large one. Just trying to keep near the top of the todos!

richardwerkman commented on src/Stryker.Core/Stryker.Core/Initialisation/ProjectInfo.cs:
```diff
@@ -8,6 +9,15 @@ namespace Stryker.Core.Initialisation
+
+        public ProjectInfo(IFileSystem fileSystem)
+        {
+            _fileSystem = fileSystem ?? new FileSystem();;
```
```suggestion
            _fileSystem = fileSystem ?? new FileSystem();
```

dupdob replied:
done

richardwerkman commented on src/Stryker.Core/Stryker.Core/MutationTest/MutationTestExecutor.cs:
```diff
@@ -70,40 +72,34 @@ public void Test(IList<Mutant> mutantsToTest, int timeoutMs, TestUpdateHandler u
-                    {
-                        result.Merge(testRunResult);
-                    }
+                    var localResult = TestRunner.TestMultipleMutants(timeoutMs, new[] {mutant}, updateHandler);
```
```suggestion
                    var localResult = TestRunner.TestMultipleMutants(timeoutMs, new[] { mutant }, updateHandler);
```

richardwerkman commented on src/Stryker.Core/Stryker.Core/MutationTest/MutationTestExecutor.cs:
```diff
@@ -70,40 +72,34 @@ public void Test(IList<Mutant> mutantsToTest, int timeoutMs, TestUpdateHandler u
-                result = multi.TestMultipleMutants(timeoutMs, mutantsToTest.ToList(), updateHandler);
-            }
-            else
+            if (forceSingle && mutantsToTest.Count>1)
```
```suggestion
            if (forceSingle && mutantsToTest.Count > 1)
```

dupdob replied:
done

richardwerkman commented on src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs:
```diff
@@ -175,23 +188,28 @@ private IEnumerable<List<Mutant>> BuildMutantGroupsForTest(IReadOnlyCollection<M
             for (var i = 0; i < mutantsToGroup.Count; i++)
             {
-                var usedTests = mutantsToGroup[i].CoveringTests.GetList().ToList();
+                var usedTests =  mutantsToGroup[i].CoveringTests;
```
```suggestion
                var usedTests = mutantsToGroup[i].CoveringTests;
```

dupdob replied:
done

richardwerkman commented on src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:
```diff
@@ -262,25 +306,46 @@ private void ParseResultsForCoverage(IEnumerable<TestResult> testResults, IEnume
             // since we analyze mutant coverage, mutants are assumed as not covered
             var seenTestCases = new HashSet<Guid>();
             var dynamicTestCases = new HashSet<Guid>();
+            var mutantCount = mutants.Max( m=> m.Id)+1;
```
```suggestion
            var mutantCount = mutants.Max(m => m.Id) + 1;
```

dupdob replied:
done

richardwerkman commented on src/Stryker.Core/Stryker.Core/TestRunners/VsTest/VsTestRunner.cs:
```diff
@@ -418,8 +506,7 @@ private string GenerateRunSettings(int? timeout, bool forMutantTesting, bool for
 {settingsForCoverage}
-<DesignMode>false</DesignMode>
-<BatchSize>1</BatchSize>
+<DesignMode>false</DesignMode><BatchSize>1</BatchSize>
```
```suggestion
<DesignMode>false</DesignMode>
<BatchSize>1</BatchSize>
```

dupdob replied:
done

richardwerkman commented on src/Stryker.DataCollector/Stryker.DataCollector/CoverageCollector.cs:
```diff
@@ -158,12 +167,16 @@ private void ReadConfiguration(string configuration)
                         }
                     }
                 }
+                if (mutations.Count==1)
```
```suggestion
                if (mutations.Count == 1)
```

dupdob replied:
done

richardwerkman reviewed (changes requested):
Sorry for this load of comments. Most are just formatting. Overall I'm impressed with the quality of your work! Thanks once again for this amazing contribution :)

dupdob commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+
+#### Run time theories
+Run time theories are discovered as one test, disregarding the number of underlying data set.
+There is one test result per dataset, all associated with the unique test id.
```
the phrase is ambiguous, and its trickier than that. I'll propose another version

dupdob commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+At coverage capture it is a problem, because there is a risk of spilling coverage information to the next tests.
+And during execution phase, it is impossible to predict when the test will really be over. 
+
+Also, if a mutation ends up changing a test case name - typically by changing the result of `ToTring()`, this
```
I tried to clarify it a bit more.

dupdob commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -0,0 +1,114 @@
+whereas it is likely these mutations would need to be adequately associated with their realtive test case.
+3. TestCaseSource is called **before** Testcase start event which means Stryker may set the active mutation too late. 
+The only way to handle this is to force any concerned mutants to run in dedicated test sessions.
+
```
I have not looked into MsTest, but
MsTest is not the most use framework and I assume there is a good integration between MsTest and VsTest. In any case, the remediation strategies I used to work around those issues should also work for MsTest based assemblies.
And, I am not sure this document is of any use, beyond technical curiosity. 
Completing means adding a new test project using MsTest and intense debugging to understand how it works.
I think It can wait for a rainy day :-).

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/Compiling/RollbackProcessTests.cs:
```diff
@@ -678,13 +679,9 @@ public void Break()
     using (var ms = new MemoryStream())
     {
-        var fixedCompilation = target.Start(compiler, compiler.Emit(ms).Diagnostics, false,false);
+                var fixedCompilation = target.Start(compiler, compiler.Emit(ms).Diagnostics, false,false);
```
Looks like an accident
fixed

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InitialTestProcessTests.cs:
```diff
@@ -7,43 +7,71 @@
             var testRunnerMock = new Mock<ITestRunner>(MockBehavior.Strict);
-            testRunnerMock.Setup(x => x.RunAll(It.IsAny<int?>(), null, null)).Returns(new TestRunResult(false) );
-            testRunnerMock.Setup(x => x.CaptureCoverage( It.IsAny<List<Mutant>>(),false, false))
+            var test1 = Guid.NewGuid();
```
improved

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/ProjectOrchestratorTests.cs:
```diff
@@ -31,15 +31,9 @@ public ProjectOrchestratorTests()
             _projectMutatorMock.Setup(x => x.MutateProject(It.IsAny<IStrykerOptions>(), It.IsAny<IReporter>()))
                 .Returns(new Mock<IMutationTestProcess>().Object);
 
-            var components = new Collection<IProjectComponent>() {
```
Don't know: there is no reference to this in my git history, so I guess it was copy pasted stuff...
not used  ⇒ removed

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/TimeoutValueCalculatorTests.cs:
```diff
@@ -1,39 +1,20 @@
+        [InlineData(1000, 0, 1500)]
+        [InlineData(1000, 2000, 3500)]
+        [InlineData(4000, 2000, 8000)]
+        public void Calculator_ShouldCalculateTimeoutValueNoExtra(int baseTime, int extra, int expected)
```
thanks

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/MutantFilters/DiffMutantFilterTests.cs:
```diff
@@ -97,34 +89,40 @@ public void ShouldOnlyMutateChangedFiles()
             // If a file inside the test project is changed, a test has been changed
-            string myTest = Path.Combine(testProjectPath, "myTest.cs"); ;
-            diffProvider.Setup(x => x.ScanDiff()).Returns(new DiffResult()
+            var myTest = Path.Combine(testProjectPath, "myTest.cs"); ;
```
done

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/MutationTest/FullRunScenario.cs:
```diff
@@ -0,0 +1,184 @@
+        public IDictionary<int, Mutant> Mutants => _mutants;
+
+        /// <summary>
+        /// Crearte a mutant
```
nice type

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/MutationTest/FullRunScenario.cs:
```diff
@@ -0,0 +1,184 @@
+                return list;
+            }
+
+            if (id == InitialRunID)
```
added a comment: the initial test run requires to return all test

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/MutationTest/MutationTestProcessTests.cs:
```diff
@@ -305,30 +303,155 @@ public void ShouldCallExecutorForEveryMutant()
+                        { "TargetDir", "/bin/Debug/netcoreapp2.1" },
+                        { "TargetFileName", "TestName.dll" },
+                        { "Language", "C#" }
+                    }).Object
+                    ,
```
fixed

dupdob commented on src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnersShould.cs:
```diff
@@ -45,6 +46,7 @@ public class VsTestRunnersShould
+        private readonly TestProperty _unexpectedCoverageProperty;
 
         // initialize the test context and mock objects
         public VsTestRunnersShould()
```
done

dupdob commented on src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs:
```diff
@@ -3,55 +3,79 @@
 
 namespace Stryker.Core.Initialisation
 {
+    public class InitialTestRun
```
done

dupdob commented on src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs:
```diff
@@ -3,55 +3,79 @@
+        public InitialTestRun InitialTest(IStrykerOptions options, ITestRunner testRunner)
         {
-            var message = testRunner.DiscoverNumberOfTests() is var total && total == -1 ? "Unable to detect" : $"{total}";
+            var message = testRunner.DiscoverTests() is var total && total.Count == 0 ? "Unable to detect" : $"{total.Count}";
```
done

dupdob commented on src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs:
```diff
@@ -3,55 +3,79 @@
 
-            return (int)stopwatch.ElapsedMilliseconds;
+            TimeoutValueCalculator = new TimeoutValueCalculator(options.AdditionalTimeoutMS,
+                (int)stopwatch.ElapsedMilliseconds-(int)_initTestRunResult.Duration.TotalMilliseconds ,
```
done

dupdob commented on src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs:
```diff
@@ -3,55 +3,79 @@
-                throw new StrykerInputException("Initial testrun was not successful.", testResult.ResultMessage);
+                var failingTestsCount = _initTestRunResult.FailingTests.Count;
+                _logger.LogWarning($"{failingTestsCount} tests are failing.");
+                if (((double)failingTestsCount) / _initTestRunResult.RanTests.Count > .1)
```
Technically, we need just one test, but if several tests are failing, this is something that could be fixed before dealing with mutants. 
More often than not, if you many failing tests, this is a sure sign that something went wrong during the build or test phase.
So I feel this should be a low tolerance by default, but it could configurable within the new option system, I guess.

but ultimately, this could be any number you are confortable with

richardwerkman replied:
I'd like to see at least 50%. We don't know the context, why some tests are failing. We print the warning so they can think for themselves if they want to continue or break up the run.

dupdob commented on src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs:
```diff
@@ -3,55 +3,79 @@
-                _logger.LogWarning("Initial test run failed. Mutation score cannot be computed.");
-                throw new StrykerInputException("Initial testrun was not successful.", testResult.ResultMessage);
+                var failingTestsCount = _initTestRunResult.FailingTests.Count;
+                _logger.LogWarning($"{failingTestsCount} tests are failing.");
```
done

dupdob commented:
Just pushed a couple of commits:
1. updated the PR to match V1 HEAD
2. added static marking for static fields and properties initial values as a likely fix #1606.

I consider the fix as 'likely' as I did not go through the effort of replicating it.

richardwerkman commented on docs/technical-reference/Testing Framework.md:
```diff
@@ -51,18 +51,22 @@ TestCase end event
+And during execution phase, it is impossible to predict when the test will really be over, so it is diffult to
+establish is the test was succesful.  
 
+Also, if a mutation ends up changing a test case name - typically by changing the result of `ToTring()`, it will change the
```
```suggestion
Also, if a mutation ends up changing a test case name - typically by changing the result of `ToString()`, it will change the
```

richardwerkman replied:
I've never thought about this. Is it an idea to skip mutating ToString methods to prevent this?

dupdob replied:
1) I enclosed detection logic that ensure any such offending mutants will be run against all tests
2) it can only happens for theories, which raises many concerns about testing in general

So, I think we don't need it, and its not worth it :-)

richardwerkman commented on src/Stryker.Core/Stryker.Core.UnitTest/Initialisation/InitialTestProcessTests.cs:
```diff
@@ -27,10 +27,10 @@ public InitialTestProcessTests()
-            var failedTests = new TestsGuidList(new[] {test1});
+            var failedTest = Guid.NewGuid();
+            var successfulTest = Guid.NewGuid();
+            var ranTests = new TestsGuidList(new[] {failedTest, successfulTest});
```
I'm now thinking, maybe we could add a constructor for TestsGuidList that takes a params param? That way we don't need to create a array for each list.

dupdob replied:
done, but it turns out it is only used 4 times. Maybe it will be helpful in the future

richardwerkman commented on src/Stryker.Core/Stryker.Core.UnitTest/Mutants/MutantPlacerTests.cs:
```diff
@@ -42,19 +42,18 @@ public void MutantPlacer_ShouldPlaceWithIfStatement(int id)
-                actualNode.DescendantNodes().First(t => t is T) as T;
+            node = actualNode.DescendantNodes().First(t => t is T {ContainsAnnotations: true}) as T;
             // Remove helper
             var restored= MutantPlacer.RemoveMutant(node);
```
```suggestion
            var restored = MutantPlacer.RemoveMutant(node);
```

richardwerkman commented on src/Stryker.Core/Stryker.Core/Initialisation/InitialTestProcess.cs:
```diff
@@ -64,15 +51,15 @@ public InitialTestRun InitialTest(IStrykerOptions options, ITestRunner testRunne
                 var failingTestsCount = _initTestRunResult.FailingTests.Count;
-                _logger.LogWarning($"{failingTestsCount} tests are failing.");
+                _logger.LogWarning($"{failingTestsCount} tests are failing. Stryker will continue but outcome will be impacted.");
                 if (((double)failingTestsCount) / _initTestRunResult.RanTests.Count > .1)
```
```suggestion
                if (((double)failingTestsCount) / _initTestRunResult.RanTests.Count >= .5)
```

richardwerkman commented on src/Stryker.Core/Stryker.Core/InjectedHelpers/CodeInjection.cs:
```diff
@@ -50,16 +52,37 @@ private static string GetRandomNamespace()
+
+        public static bool IsContextAccessExpression(ExpressionSyntax memberAccess, string member) =>
+            memberAccess is MemberAccessExpressionSyntax
+            {Expression: MemberAccessExpressionSyntax
```
This method can use some outlining. It's a bit messy now

dupdob replied:
done, hopes it looks better now. I don't know the rules for pattern matching formatting, I am afraid

richardwerkman commented on src/Stryker.Core/Stryker.Core/Instrumentation/StaticInitializerMarkerEngine.cs:
```diff
@@ -0,0 +1,42 @@
+        {
+        }
+
+        public ExpressionSyntax PlaceValueMarker(ExpressionSyntax node)
```
Might be nice to add a xml comment above of how the created code might look like. Reads a lot easier than this syntax factory code

dupdob replied:
done

richardwerkman commented on src/Stryker.Core/Stryker.Core.UnitTest/Mutants/MutantPlacerTests.cs:
```diff
@@ -143,6 +142,16 @@ public void ShouldInjectInitalizersAndRestore()
         }
 
+        [Fact]
+        public void ShouldStaticMarkerInStaticFieldInitializers()
```
```suggestion
        public void ShouldPlaceStaticMarkerInStaticFieldInitializers()
```

richardwerkman commented on src/Stryker.Core/Stryker.Core/Instrumentation/StaticInitializerMarkerEngine.cs:
```diff
@@ -0,0 +1,42 @@
+
+namespace Stryker.Core.Instrumentation
+{
+    internal class StaticInitializerMarkerEngine: BaseEngine<ExpressionSyntax>
```
I'm missing tests for this engine?

dupdob replied:
no, the test is in MutantPlacerTests class

richardwerkman reviewed (changes requested):
Great additions. Still some small remarks but we're close to merging now I feel
